wsgi_kerberos.py

```diff
@@ -49,37 +56,47 @@ def _consume_request(environ, read_max):
     '''
     if read_max == 0:  # Short-circuit early when user opts out of this.
         return
+    if environ["REQUEST_METHOD"] in _NEVER_READ_METHODS:
+        return
     if environ.get("HTTP_EXPECT") == "100-continue":
         return
     try:
-        sock = environ.get('wsgi.input')
-        if hasattr(sock, 'closed') and sock.closed:
+        body = environ.get('wsgi.input')
+        if hasattr(body, 'closed') and body.closed:
             return
 
         content_length = environ.get('CONTENT_LENGTH', '')
         if not content_length:
-            remaining = read_max
-        else:
-            content_length = int(content_length)
-            if content_length > read_max:
-                # User is not willing to read such a large request body, but
-                # reading anything less does not help naively-written clients.
-                return
-            remaining = content_length
-
+            LOG.debug("No Content-Length -> skipping _consume_request")
+            return
+        content_length = int(content_length)
+        if content_length > read_max:
+            # Server is not willing to read such a large request body, but
+            # reading anything less does not help naively-written clients.
+            LOG.warning(
+                "Content-Length (%s) exceeds read_max (%s) -> not consuming"
+                " request body; client may get a connection error. Enabling"
+                " preemptive authentication on the client may avoid this."
+                " You can also pass a higher value of `read_max_on_auth_fail`"
+                " to KerberosAuthMiddleware.",
+                content_length, read_max
+            )
+            return
         # Try to receive all of the data. Keep retrying until we get an error
         # which indicates that we can't retry. Eat errors. The client will just
         # have to deal with a possible Broken Pipe -- we tried.
+        remaining = content_length
         while remaining > 0:
+            to_read = min(io.DEFAULT_BUFFER_SIZE, remaining)
             try:
-                delta = len(sock.read(remaining))
+                read = len(body.read(to_read))
             except socket.error as err:
                 if err.errno != errno.EAGAIN:
-                    break
+                    raise
             else:
-                if delta == 0:
+                if read != to_read:
                     break
```

Comment:
Since `environ['wsgi.input']` is actually a file-like object rather than a raw socket (which is now made clearer by renaming `sock` to `body` above), I realized that the only time `read` will read less than the number of bytes we requested is when the client disconnects before sending all the data that it said it would. So our read loop should terminate not only when an iteration reads 0 bytes, but when it reads anything other than the amount requested.

This better matches Werkzeug's `LimitedStream.exhaust()` implementation (and thanks for reminding me of that!):
https://github.com/pallets/werkzeug/blob/f3b081634c744ae1e0ae3f76034ee90bfdc2a808/src/werkzeug/wsgi.py#L972

Reply:
Sounds good